File: src/cpp/config/backpackConfig.cpp
/*
*	backpackConfig.cpp
*
*	This class provides an easy method of importing and accessing the
*	hardware configuration numbers from a particular revision of the 
*	backpack.  This class expects to be initialized with an external
*	.xml file that holds the information
*/

#include "backpackConfig.h"

// Includes
#include <vector>
#include <string>
#include <map>
#include <iostream>
#include <ostream>
#include <config/sensorProp.h>
#include <xmlreader/tinyxml.h>
#include <xmlreader/tinyxmltools.h>

// Namespace
using namespace std;

/*
	backpackConfig(std::ostream& logger = std::cout);
	backpackConfig(std::string config_file_name, std::ostream& logger = std::cout);

	Constructors.  Either an empty class can be constructed or a config XML 
	can be read into the class.
*/
backpackConfig::backpackConfig(std::ostream& logger) {
	this->clear();
	this->log_stream = &logger;
}
backpackConfig::backpackConfig(std::string config_file_name,
							   std::ostream& logger) {
	this->log_stream = &logger;
	if( !this->read_config_file(config_file_name) ) {
		this->clear();
	}
}

/*
	void clear()

	Clears the class data members to the state when they were created with
	the default constructor
*/
void backpackConfig::clear() {

	// Maps from label names to pointers
	this->pointer_map.clear();
	this->instance_map.clear();

	// Set to not read state
	this->is_read = false;

}

/*
	bool read_config_file(std::string config_file_name);

	This function is the main function called to read a backpackConfig XML
	files data into the classes data structure elements
*/
bool backpackConfig::read_config_file(std::string config_file_name) {

	// The first thing we need to do is clear the class of any
	// data
	this->clear();

	// Next we need to open and parse the xml documenta
	if( !this->xml_doc.LoadFile(config_file_name.c_str()) ) {
		*(this->log_stream) << "[backpackConfig::read_config_file] - "
			<< "Unable to parse xml file." << endl;
		return false;
	}

	// If we were able to successfully parse and validate the xml
	// file for correct xml structure, then we need to validate
	// that this file 
	// 1) has the correct top level nametag
	// 2) that tag is unique
	//
	// First we check 1)
	TiXmlNode * top_level_ptr = this->xml_doc.FirstChild("sensors");
	if( top_level_ptr == NULL ) {
		*this->log_stream << "[backpackConfig::read_config_file] - "
			<< "Top level tag should be <sensors>!" << endl;
		return false;
	}
	// Check 2)
	if( top_level_ptr != this->xml_doc.LastChild("sensors") ) {
		*this->log_stream << "[backpackConfig::read_config_file] - "
			<< "<sensors> tag is multiply defined." << endl;
		return false;
	}
	
	// Next what we need to do is locate all sensor types that
	// are in this file
	TiXmlNode * sensor_type_ptr = top_level_ptr->FirstChildElement();
	while(sensor_type_ptr != NULL) {

		// Copy out the value name
		string type_name = sensor_type_ptr->Value();

		// Check if it is in the map already...  it shouldnt be.
		// If it is then the tag is multiply defined and we should
		// throw an error
		if( this->pointer_map.count(type_name) != 0 ) {
			*this->log_stream << "[backpackConfig::read_config_file] - "
			<< "<" << type_name << "> tag is multiply defined." << endl;
			return false;
		}

		// Insert it into the map
		this->pointer_map[type_name] = sensor_type_ptr;

		// Iterate to next
		sensor_type_ptr = sensor_type_ptr->NextSiblingElement();
	}

	// Next for each type of scanner we want to make a mapping between
	// the type names and the actual instances of the sensor
	map<string, TiXmlNode *>::iterator type_iterator;	
	for( type_iterator = this->pointer_map.begin(); 
		type_iterator != this->pointer_map.end(); type_iterator++) {

			// Make a copy for easy usage
			string type_name = type_iterator->first;
			sensor_type_ptr =  type_iterator->second;

			// Loop over the number of each instance and insert the
			// pointers into the instance_map
			TiXmlNode * instance_pointer = sensor_type_ptr->FirstChildElement();
			while(instance_pointer != NULL) {
				this->instance_map[type_name].push_back(instance_pointer);
				instance_pointer = instance_pointer->NextSiblingElement();
			}
	}

	// If we were able to do all of this successfully,
	// Then we can return success
	this->is_read = true;
	return true;
}



/*
	bool build_property_map(TiXmlNode * instance_ptr,
							std::map<std::string,std::string>& property_map);

	Builds the property map for a sensor instance.  Stores it in the property_map.
*/
bool backpackConfig::build_property_map(TiXmlNode * instance_ptr,
									    std::map<std::string,std::string>& property_map) {

	// Clear the map of any previous elements
	property_map.clear();

	// Extract the elements
	TiXmlNode * element_ptr = instance_ptr->FirstChildElement();
	while(element_ptr != NULL) {

		// Find element name
		string element_name = element_ptr->Value();

		// Check to see if we have a single text element underneath it
		if( TiXmlTools::countChildTextElements(element_ptr) != 1 ||
		    TiXmlTools::countChildNodes(element_ptr) != 1 )  {
			property_map.clear();
			return false;
		}

		// Find the element string
		string element_text = element_ptr->FirstChild()->Value();

		// Add to the map
		property_map[element_name] = element_text;

		// Move on
		element_ptr = element_ptr->NextSiblingElement();
	}

	

	// Return Success
	return true;
}

File: src/python/config/backpackconfig.py
#
# 	backpackconfig.py
#
#	This module gives an interface for working with backpack config xml files
#
#	It allows a user to easily travel through the xml hierarchy associated
#	with a tree of .xml config files
#

import os
import sys
import xml.etree.ElementTree as ET

class Configuration :

	#
	#	def __init__(self, filename, trimByEnable=True, relativeDir=None)
	#
	# 	The constructor for the configuration file
	#	parses the tree of xml files importing all of the settings
	# 	into a single element tree
	#
	#	If you give relativeDir it resolves all relative paths
	# 	to that directory, if it is not given then all paths
	#	are assumed to be relative to the file they are in
	#
	#	Also optionally you can specify to trim the thing by an enable
	#	flag
	#
	def __init__(self, filename, trimByEnable=True, relativeDir=None) :

		# Records the filename used when importing
		self.filename = filename
		self.relativeDir = relativeDir

		# Set the flag for if there was an error
		#
		#	0 - No error
		#	1 - Unable to parse a file
		#	2 - Circular File Tree Detected, Branch was aborted
		#
		# 	This is actually bit packed so you need to and
		# 	the value stored here with the error code above
		#	to detect the problem case
		#
		self.wasError, self.tree = parse_files(filename, relativeDir, trimByEnable)

	#
	#	def dump()
	#
	# 	Dumps the tree for debugging
	#
	def dump(self) :
		ET.dump(self.tree)

	#
	#	def circular_file_tree()
	#
	#	Returns a boolean that indicates if the xml file tree was
	#	circularly defined
	#
	def circular_file_tree(self) :
		return (self.wasError | 2) == 2

	#
	#	def import_error(self) :
	#
	#	Returns if there was an error importing a childs
	#	xml files somewhere in the backpack xml tree
	#
	def import_error(self) :
		return (self.wasError | 1) == 1

	#
	#	def find(self, findstr)
	#
	#	Wraps ETs find functionality
	#
	def find(self, findstr) :
		return self.tree.getroot().find(findstr)

	#
	#	def findall(self, findstr)
	#
	#	Wraps ETs findall functionality
	#
	def findall(self, findstr) :
		return self.tree.getroot().findall(findstr)

	#
	#	def hardware_version(self)
	#
	#	Returns the hardware version string
	#
	def hardware_version(self) :
		hardwareVersion = self.tree.getroot().find("./hardware/version")
		if hardwareVersion != None :
			return hardwareVersion.text.strip()

	#
	#	def hardware_codename(self)
	#
	#	Returns the codename string of the hardware
	#
	def hardware_codename(self) :
		hardwareCodename = self.tree.getroot().find("./hardware/codename")
		if hardwareCodename != None :
			return hardwareCodename.text.strip()

	#
	#	def find_sensor(self, sensorName, sensorType="*")
	#
	#	Finds a specific sensors tree based on a type and a sensor name.
	#	A type can be specified to speed up the search or resolve 
	#	name conflicts between different sensor types
	#
	def find_sensor(self, sensorName, sensorType="*") :
		sensorList = self.tree.getroot().findall("./"+sensorType)
		for sensorType in sensorList :
			for sensor in sensorType :
				nameElement = sensor.find("./name")
				if nameElement != None :
					if nameElement.text.strip() == sensorName :
						return sensor
		return None

	# 
	#	def find_sensors_by_type(self, sensorType)
	#
	#	Returns a list of all sensors names who have a specific sensor type
	#
	def find_sensors_by_type(self, sensorType) :
		sensorList = self.tree.getroot().findall("./"+sensorType)
		retList = []
		for sensorType in sensorList :
			for sensor in sensorType :
				nameElement = sensor.find("./name")
				if nameElement != None :
					retList.append(nameElement.text.strip())

		# Return the list
		return retList

	#
	#	def find_sensor_prop(self, sensorName, sensorProp, sensorType="*")
	#
	#	Looks up a sensor property based on sensor name.
	#
	#	sensorName  -  	specifies the name of the sensor from the configuration
	#				   	file
	#	sensorProp 	- 	what property you want to look for.  This can be a path
	#					like object such as "configFile/fps" for properties that
	#					are nested more levels down
	#	sensorType	-	Specifies what type of sensor you are looking for to cut
	#					down on search time or to differentiate between different
	#					types of sensor with the same name.
	#	
	#	Returns None if the property is not found 
	#
	def find_sensor_prop(self, sensorName, sensorProp, sensorType="*") :
		
		# Find the sensor
		sensor = self.find_sensor(sensorName, sensorType)
		if sensor == None :
			return None

		# Find the property
		prop = sensor.find("./"+sensorProp)
		if prop == None :
			return None

		# Return the text
		if prop.text == None :
			return None
		return prop.text.strip()


#
# def parse_files(filename, relativeDir, trimByEnable)
#
# Runs the actual import code
#
def parse_files(filename, relativeDir, trimByEnable) :

	# First we need to open the initial file
	try :
		tree = ET.parse(filename)
	except :
		return 1, None

	# If we want to trim by enable we do this now
	if trimByEnable :
		trim_by_enable(tree.getroot())

	# Next we do the actual imports
	filename = os.path.abspath(filename)
	seenFiles = [filename]
	error, root = recursive_import(tree.getroot(), seenFiles, 0, filename, relativeDir)

	# Return the tree and the error status
	return error, tree

#
#	def recursive_import(thisElem, seenFiles. lastError, parentFile, relativeDir)
#
#	The actual recursive part
#
def recursive_import(thisElem, seenFiles, lastError, parentFile, relativeDir) :

	# The first thing we do is check if this particular element
	# needs has an xml file associated with it
	thisText = thisElem.text
	if thisText == None :
		thisExt = ""
	else :
		thisText = thisText.strip()
		thisFile, thisExt = os.path.splitext(thisText)

	if thisExt.lower() == '.xml' :

		# Attempt to import the xml file
		try :
			
			# Filepaths are assumed to be relative, so we tack
			# it onto the parentFiles path
			if relativeDir == None :
				head, tail = os.path.split(parentFile)
			else :
				head = relativeDir
			newParentFile = os.path.normpath(os.path.join(head, thisFile+thisExt))

			newRoot = ET.parse(newParentFile).getroot()

			# Record that we saw this file if we have not seen it already
			if newParentFile in seenFiles :
				return (lastError | 2), thisElem
			else :
				seenFiles.insert(0, newParentFile)

			# Insert this into the tree at the current place
			clone_tree(newRoot, thisElem)

			# Recurse over this subtree
			lastError, thisElem = recursive_import(newRoot, seenFiles, lastError, newParentFile, relativeDir)

		except Exception as e:
			return (lastError | 1), thisElem

	# Iterate the child elements of this element looking for other xml files
	for child in thisElem :
		lastError, thisElem = recursive_import(child, seenFiles, lastError, parentFile, relativeDir)

	return lastError, thisElem

#
#	def clone_tree(inroot, outroot)
#
def clone_tree(inroot, outroot) :

	for child in inroot :
		newOutRoot = ET.SubElement(outroot, child.tag, child.attrib)
		newOutRoot.text = child.text
		clone_tree(child, newOutRoot)

#
#	def trim_by_enable(root)
#
#	Trims the tree of any nodes and subnodes if they
#	contain a tag that is "enable" and it is set to
#	0
#
def trim_by_enable(root) :

	# Check if the child has an enable flag
	toRemove = []
	for child in root :
		enableElem = child.find("./enable")
		if enableElem != None :
			
			# if it is not enabled flag it for descruction it
			isEnabled = (int(enableElem.text) != 0) 
			if not isEnabled :
				toRemove.append(child)

	# Destory things to remove
	for child in toRemove :
		root.remove(child)

	# Recurse
	for child in root :
		trim_by_enable(child)



